### DISCUSSION TITLE
mesh_tri.isValid()

### DISCUSSION THREAD

**Discussion by JarosT55** [kinnala/scikit-fem#1102] [Q&A] [answered]
**Hi,**
I accoutered the problem in the following sequence of the code:

$meshT = MeshTri().refined(2)$
$isValid = meshT.is⎽valid()$

produces:   $\quad\quad\quad isValid = False$

The problem stems from the comparison inside $is⎽valid()$ .
$mesh.doflocs.shape[0] != mesh.elem.refdom.dim()\quad\quad\quad$                **# 2 != 3**

The similar code for $MeshQuad()$ does not exhibit any issues.

I looked inside **_.skfem\refdom.py_**, but the finesse of what was going on there is **beyond my rusty comprehension.**

Any ideas ...? 

**Regards
MichaelT**

**Comment by kinnala** [comment, marked as answer]
I am unable to reproduce the issue:
```python
In [1]: from skfem import *

In [2]: MeshTri().refined(2).is_valid()
Out[2]: True
```
I am using the latest main branch.

**Comment by JarosT55** [comment]
**Hi Mr Kinnala,**

**Thank you** for checking out.
I updated to the latest version of **skfem** 
_Successfully uninstalled scikit-fem-8.1.0
Successfully installed scikit-fem-9.0.1_
and I also re-checked in the minimalist code **in the separate Jupyter notebook**, 
The result is positive—no validity problem … **for vanilla-type mesh**.
The problem existed in the original code and only when it was run for **the second and successive times**. **Import/dealings with 3D surface triangular meshes triggered it.**

As I understand, **skfem** does not support such **3D surface meshes** and respective capable elements.
Can they be ‘_simulated_’ by ‘_flattened_’ **ElementTet()s**? _(rather crazy idea!)_

**Regards
MichaelT**

**Reply by kinnala** [reply]
Yes they can, it is not a crazy idea, I have used it a few times. It is currently the only way to use such surface meshes in scikit-fem. I have tried to support them natively once or twice but they seem to require so many changes here and there that I have given up (due to lack of motivation - I don't study such problems in my own research).

**Reply by kinnala** [reply]
If you read a surface mesh, you should add to each element a fourth node to make it a tetrahedron. Then in the final matrices you should slice only the nodes belonging to the surface. That way you can use scikit-fem for surface meshes.

**Reply by JarosT55** [reply]
**Hi Mr Kinnala,** 

... I was thinking along this line but _in the opposite direction_, ... by using a **sledgehammer to flatten tetrahedra** :).
Although possible, full implementation of a **3D surface** in **skfem** would be pretty challenging. Nevertheless, the package's fabric and its **good structure** seems suitable for such a task.
I think I will try finding other '_countermeasures_' to deal with **3D surface meshes**, … anyway, only for visualization purposes.
**Thank you** for the input.

**Regards
MichaelT**

**Reply by JarosT55** [reply]
**Hi Mr Kinnala,** 

I have dealt … a little with **'3D surface'** issue. Despite some progress, I encountered a **'topology'** problem.
Perhaps, it deserves, in its own right, the separate discussion thread.
I will start the new one under the title **'Mesh and Normals'** ⏳.

**Regards
MichaelT**
